<?xml version="1.0" encoding="UTF-8"?>

<launch>

  <!-- Arguments -->
  <arg name="launch_prefix" default="" doc="gdb -ex run --args"/>
  <arg name="launch_rviz" default="true"/>
  <arg name="launch_tf" default="false"/>
  <arg name="publish_gps_tf" default="n" doc="y , n"/>
  <arg name="load_params" default="n" doc="y , n"/>
  <arg name="robot_description_file_path" default="/home/tutuna/Music/hongg_wildcat_mission_4/2022-12-15-13-52-33_anymal-camel-lpc_mission.yaml"/>
  
  <!-- Common Options:  -->
  <!--        /point_cloud_filter/lidar_depth_camera/point_cloud_filtered   -->
  <!--        /lidar/point_cloud  -->
  <!--        /point_cloud_filter/lidar/point_cloud_filtered -->
  <!--        /hesai/pandar  -->
  <arg name="cloud_topic" default="/point_cloud_filter/lidar_depth_camera/point_cloud_filtered"/> <!-- /hesai/pandar /point_cloud_filter/lidar_depth_camera/point_cloud_filtered  /lidar/point_cloud -->
  
   <!-- These 3 topics are exclusive. Only one of them going to be used even if multiple exist. This is done to support multiple msg types. -->
  <arg name="odometry_topic" default="/graph_msf/est_odometry_odom_imu2"/> <!-- /point_cloud_filter/lidar_depth_camera/odometry_pose -->
  <arg name="pose_stamped_topic" default="/point_cloud_filter/lidar_depth_camera/odometry_pose_noised2"/> <!-- /point_cloud_filter/lidar_depth_camera/odometry_pose -->
  <arg name="pose_stamped_with_covariance_topic" default="/state_estimator/pose_in_odom2"/> <!-- /point_cloud_filter/lidar_depth_camera/odometry_pose -->

  <arg name="parameter_filename" default="param_velodyne_puck16_replay_only_x.lua"/>
  <arg name="parameter_folder_path" default="$(find open3d_slam_ros)/param/"/>
  <arg name="map_saving_folder" default="$(find open3d_slam_ros)/data/maps/"/>
  <arg name="num_accumulated_range_data" default="1"/>
  
  <!-- Replay Options -->
  <arg name="is_read_from_rosbag" default="true"/>
  <!-- /home/tutuna/Music/only_x_analysis_dataset/revision_very_noisy_request/merged_high_noise.bag -->
  <!-- /home/tutuna/Music/only_x_analysis_dataset/merged.bag -->
  <arg name="rosbag_filepath" default="/home/tutuna/Music/only_x_analysis_dataset/revision_very_noisy_request/merged_high_noise.bag"/> <!-- /home/tutuna/Downloads/mergedBag.bag -->
  <arg name="use_syncronized_poses_to_replay" default="false"/>
  <arg name="republish_tf_topic" default="false"/>
  <arg name="async_pose_topic" default="/state_estimator/pose_in_odom"/> <!-- /state_estimator/pose_in_odom  /noised_prior  /very_noised_prior -->
  <arg name="relative_sleep_duration" default="0.0"/>
  <arg name="replay_start_time_as_second" default="0.0"/> <!-- 1440 for tunnel-->
  <arg name="replay_end_time_as_second" default="5170.0"/> <!-- 380  for hongg anymal  2170 for tunnel -->
  <arg name="use_sim_time" default="true"/>
  <arg name="assumed_external_odometry_tracked_frame" default="base"/> <!-- imu_box_link for julian exca zed2i_base_link This is used to find the transformation to the "tracked_sensor_frame" -->
  <arg name="export_imu_data" default="false"/> <!-- This is used to export the IMU data from the rosbag. Also sleeps 10s. -->
  <arg name="use_gps_for_ground_truth" default="false"/>
  <arg name="gps_frame" default="gps_mock"/> <!-- gps_mock for hongg in music -->
  <arg name="gps_topic" default="/rtk_gps_driver/position_receiver_0/ros/navsatfix"/>
  <arg name="downsample_cloud_for_replay" default="false"/>    
  <arg name="save_noised_prior" default="false"/>
  <arg name="save_processed_bag" default="false"/>
  <arg name="downSamplingSkippingRate_" default="10"/>
  <arg name="icp_yaml_full_path" default="$(find open3d_slam_ros)/param/icp.yaml"/>   


  <!-- ROS Param -->
  <param name="/use_sim_time" value="$(arg use_sim_time)"/>

  <!-- Main Node -->
  <node name="open3d_slam" pkg="open3d_slam_ros"
        type="mapping_node" output="screen"
        launch-prefix="$(arg launch_prefix)">
    <param name="cloud_topic" type="string" value="$(arg cloud_topic)"/>

    <param name="odometry_topic" type="string" value="$(arg odometry_topic)"/>
    <param name="pose_stamped_topic" type="string" value="$(arg pose_stamped_topic)"/>
    <param name="pose_stamped_with_covariance_topic" type="string" value="$(arg pose_stamped_with_covariance_topic)"/>

    <param name="parameter_folder_path" type="string"
           value="$(arg parameter_folder_path)"/>

    <param name="assumed_external_odometry_tracked_frame" type="string"
           value="$(arg assumed_external_odometry_tracked_frame)"/>

    <param name="async_pose_topic" type="string"
           value="$(arg async_pose_topic)"/>

    <param name="export_imu_data" type="bool"
           value="$(arg export_imu_data)"/>

    <param name="gps_frame" type="string"
           value="$(arg gps_frame)"/>

    <param name="use_gps_for_ground_truth" type="bool"
           value="$(arg use_gps_for_ground_truth)"/>           

    <param name="downsample_cloud_for_replay" type="bool"
           value="$(arg downsample_cloud_for_replay)"/>  

    <param name="save_noised_prior" type="bool"
           value="$(arg save_noised_prior)"/>

    <param name="save_processed_bag" type="bool"
           value="$(arg save_processed_bag)"/>

    <param name="downSamplingSkippingRate_" type="int"
           value="$(arg downSamplingSkippingRate_)"/>  

    <param name="icp_yaml_full_path" type="string"
           value="$(arg icp_yaml_full_path)"/>  

    <param name="gps_topic" type="string"
           value="$(arg gps_topic)"/>

    <param name="use_syncronized_poses_to_replay" type="bool"
           value="$(arg use_syncronized_poses_to_replay)"/>

    <param name="republish_tf_topic" type="bool"
           value="$(arg republish_tf_topic)"/>

    <param name="relative_sleep_duration" type="double"
           value="$(arg relative_sleep_duration)"/>

    <param name="replay_start_time_as_second" type="double"
           value="$(arg replay_start_time_as_second)"/>

    <param name="replay_end_time_as_second" type="double"
           value="$(arg replay_end_time_as_second)"/>

    <param name="parameter_filename" type="string"
           value="$(arg parameter_filename)"/>
    <param name="num_accumulated_range_data"
           value="$(arg num_accumulated_range_data)"/>
    <param name="is_read_from_rosbag" value="$(arg is_read_from_rosbag)"/>
    <param name="rosbag_filepath" value="$(arg rosbag_filepath)"/>
    <param name="map_saving_folder" value="$(arg map_saving_folder)"/>
  </node>

  <!-- Visualization -->
  <include
      file="$(find open3d_slam_ros)/launch/vis.launch"
      if="$(arg launch_rviz)">
  </include>

  <!-- Robot Description & Parameter Loading -->
  <group if="$(eval load_params == 'y')">
       <rosparam command="load" file="$(arg robot_description_file_path)" />
  </group>

    <!-- Anymal tf publisher, you are responsible to load the robot description. -->
  <include 
       file="$(find anymal_tf_publisher)/launch/anymal_tf_publisher.launch"
       if="$(arg launch_tf)">
  </include>

</launch>